Add CodeQL Workflow for Code Security Analysis

b4yuan opened the pull request:
# Summary

This pull request introduces a CodeQL workflow to enhance the security analysis of this repository.

# What is CodeQL

CodeQL is a static analysis tool that helps identify and mitigate security vulnerabilities. It is primarily intra-function but does provide some support for inter-function analysis. By integrating CodeQL into a GitHub Actions workflow, it can proactively identify and address potential issues before they become security threats.

For more information on CodeQL and how to interpret its results, refer to the GitHub documentation and the CodeQL documentation (https://codeql.github.com/ and https://codeql.github.com/docs/).

# What this PR does

We added a new CodeQL workflow file (.github/workflows/codeql.yml) that
- Runs on every pull request (functionality to run on every push to main branches is included as a comment for convenience).
- Runs daily.
- Excludes queries with a high false positive rate or low-severity findings.
- Does not display results for git submodules, focusing only on our own codebase.

## Validation

To validate the functionality of this workflow, we have run several test scans on the codebase and reviewed the results. The workflow successfully compiles the project, identifies issues, and provides actionable insights while reducing noise by excluding certain queries and third-party code.

## Using the workflow results

If this pull request is merged, the CodeQL workflow will be automatically run on every push to the main branch and on every pull request to the main branch. To view the results of these code scans, follow these steps:
1. Under the repository name, click on the Security tab.
2. In the left sidebar, click Code scanning alerts.

# Is this a good idea?

We are researchers at Purdue University in the USA. We are studying the potential benefits and costs of using CodeQL on open-source repositories of embedded software.

We wrote up a report of our findings so far. The TL;DR is “CodeQL outperforms the other freely-available static analysis tools, with fairly low false positive rates and lots of real defects”. You can read about the report here: https://arxiv.org/abs/2310.00205

# Review of engineering hazards

License: see the license at https://github.com/github/codeql-cli-binaries/blob/main/LICENSE.md: 

```
Here's what you may also do with the Software, but only with an Open Source Codebase and subject to the License Restrictions provisions below:

Perform analysis on the Open Source Codebase.

If the Open Source Codebase is hosted and maintained on GitHub.com, generate CodeQL databases for or during automated analysis, CI, or CD.
```

False positives: We find that around 20% of errors are false positives, but that these FPs are polarized and only a few rules contribute to most FPs. We find that the top rules contributing to FPs are: cpp/uninitialized-local, cpp/missing-check-scanf, cpp/suspicious-pointer-scaling, cpp/unbounded-write, cpp/constant-comparison, and cpp/inconsistent-null-check. Adding a filter to filter out certain rules that contribute to a high FP rate can be done simply in the workflow file.

pjonsson commented on .github/workflows/codeql-buildscript.sh:
```diff
@@ -0,0 +1,12 @@
+#!/usr/bin/env bash
```
Can the job use `tests/compile-all/build.sh` for the functionality of this script?

(That script should also stop running "make -j" without any upper limit of parallelism, but that isn't your problem.)

pjonsson replied:
And just be aware that the `tests/compile-all/build.sh` takes a while, output from `time -p make` in that directory:
```
real 2043.81
user 13620.13
sys 6830.35
```

b4yuan replied:
Yes, the job can be subbed out with any script.

pjonsson replied:
You can replace this script with a call to `make -j<cores> -C tests/compile-all` now that #2747 is merged.

If you manage to run this job inside the Docker container, we have an alias called `cimake` so you can get an even better ccache hit rate by running `cimake -j<cores> -C tests/compile-all` instead.

pjonsson commented on .github/workflows/codeql.yml:
```diff
@@ -0,0 +1,127 @@
+    - name: Checkout repository
+      uses: actions/checkout@v3
+      with:
+        submodules: recursive
```
We had problems with git-lfs and other things, so I think you want to copy the checkout action part from `.github/workflows/build.yml`.

pjonsson commented on .github/workflows/codeql.yml:
```diff
@@ -0,0 +1,127 @@
+      with:
+        submodules: recursive
+
+    - name: arm-none-eabi-gcc GNU Arm Embedded Toolchain
```
I'm not thrilled about getting one more set of build toolchains to keep updated. There is a Docker container that contains them all, can this job use that and have CodeQL run in that container?

pjonsson replied:
I ran into this link in another project:

https://docs.github.com/en/code-security/code-scanning/creating-an-advanced-setup-for-code-scanning/running-codeql-code-scanning-in-a-container

pjonsson commented on .github/workflows/codeql.yml:
```diff
@@ -0,0 +1,127 @@
+    #   - https://gh.io/supported-runners-and-hardware-resources
+    #   - https://gh.io/using-larger-runners
+    # Consider using larger runners for possible analysis time improvements.
+    runs-on: ${{ (matrix.language == 'swift' && 'macos-latest') || 'ubuntu-20.04' }}
```
Does CodeQL only work on Ubuntu 20.04, or could this be Ubuntu 22.04?

pjonsson commented on .github/workflows/codeql.yml:
```diff
@@ -0,0 +1,127 @@
+          -**/3rd*party/**
+          -**/external/**
+        input: ${{ steps.step1.outputs.sarif-output }}/cpp.sarif
+        output: ${{ steps.step1.outputs.sarif-output }}/cpp.sarif
```
I don't know all details about Github jobs, but can this step really output to some other step's output?

pjonsson commented on .github/workflows/codeql.yml:
```diff
@@ -0,0 +1,127 @@
+      with:
+        name: codeql-results
+        path: ${{ steps.step1.outputs.sarif-output }}
+        retention-days: 5
```
There's a red ring at the end of this file, is it missing a final newline?

pjonsson replied:
Even with a team of 10 full time developers, it can be hard to manage to investigate strange blips in nightly jobs within a week. Are the SARIF outputs big, or could they be retained for 30 days just as well?

pjonsson commented on .github/workflows/fail_on_error.py:
```diff
@@ -0,0 +1,34 @@
+                rule_level = rules_metadata[rule_index]['defaultConfiguration']['level']
+            except IndexError as e:
+                print(e, rule_index, len(rules_metadata))
+            else:
```
Asked as someone who doesn't speak Python: what does this else belong to?

pjonsson commented on .github/workflows/fail_on_error.py:
```diff
@@ -0,0 +1,34 @@
+import sys
+
+# Return whether SARIF file contains error-level results
+def codeql_sarif_contain_error(filename):
```
Most Python code in the repository is old, but since we are on Python 3 now, please add type hints to the function. (I'm guessing `..(filename: str) -> bool:`).

pjonsson commented on .github/workflows/codeql.yml:
```diff
@@ -0,0 +1,127 @@
+  # push:
+  #   branches: [ "main", "master" ]
+  schedule:
+    - cron: '0 0 * * *'
```
When people are dealing with merge requests, there is a shortage of CI runners. Which timezone does this job run at midnight on? How long has the job taken to complete in your test runs?

b4yuan commented:
@pjonsson I stumbled upon this: https://github.com/github/codeql/issues/8502

From what I've read about the documentation, it _should_ be possible, but it does seem to be 'a little more involved'. 

Upon further research -- we can remove the following files:

`codeql-buildscript.sh`: by implementing these steps into `codeql.yml` itself... still working on this on our fork, for some reason it's not as straightforward as it should be.

`fail_on_error.py`: there is a setting in https://github.com/contiki-ng/contiki-ng/settings/security_analysis
![image](https://github.com/contiki-ng/contiki-ng/assets/89487381/66b7a649-91e2-4b45-b5f7-ed3ceeaf06ef) which should force a PR workflow to fail.

If both of these scripts are removed, are you fine to leaving the CodeQL analysis as a separate action, outside of the container?

pjonsson commented:
> @pjonsson I stumbled upon this: [github/codeql#8502](https://github.com/github/codeql/issues/8502)

Contiki-NG itself is not built as part of building the image, so for the purpose of this PR, you can think of the docker image as just a frozen set of package versions.

Looking at https://docs.github.com/en/code-security/code-scanning/creating-an-advanced-setup-for-code-scanning/running-codeql-code-scanning-in-a-container#example-workflow, I get the impression they run just fine inside a container, it's just not the Contiki-NG container.

I don't think it's a strict requirement to have CodeQL run on every pull request, so if only having a nightly run makes life easier for you, let's go with that. (I'm aware of the tradeoffs from a user perspective.)

> If both of these scripts are removed, are you fine to leaving the CodeQL analysis as a separate action, outside of the container?

Since this is apparently more complicated than it should be (and what I thought), let's revise the plan:

Could you keep the `codeql-buildscript.sh` and most of the other steps you have in this PR, but run the scanning inside the codeql-container mentioned in https://docs.github.com/en/code-security/code-scanning/creating-an-advanced-setup-for-code-scanning/running-codeql-code-scanning-in-a-container#example-workflow?

If you get the CodeQL stuff working inside the codeql-container, we can take it from there and rebuild our own container using the codeql-container as base image (semi-modern versions of docker accept `--build-context ` as parameter) for more extensible scanning.

I did not intend to suggest `fail_on_error.py` should be removed, one comment I had was that you should add a type signature to the function in the file, and the second comment I had was a question for what a specific line of `else:` belonged to, the try-statement on line 23?

b4yuan commented:
I've gone ahead and added what you've asked. 

Regarding the Python script questions: I realized that I commented on your questions but they are marked as 'pending'... not sure if you can see them. Anyhow, I've added some type hints and yes, the `else: ` belongs to the `try` statement on line 23.

pjonsson commented:
I don't see the pending comments. I think you have to submit them somehow, possibly as a review.

Have to run now, but will take a look at the updated PR later.

b4yuan commented:
@pjonsson Hey, any update on the PR?

pjonsson commented:
@b4yuan sorry, have plumbers in the apartment this week so schedule took a hit. I still can't see your responses to my comments (beyond your response on the build script), can you check if you can find some submit button or similar so we can see them?

If you can't find a submit button, just paste the still relevant answers in the comment box at the bottom of the page.

Changes:

- Add the missing newline at the end of codeql.yaml
- Switch runs-on from Ubuntu 20.04 to Ubuntu 22.04
- Increase retention days for the results to 14
- Put the cron time so the scanning job runs at 4 am CET, that time is late in the US and people in Europe should definitely have stopped working at that time, so CI runners should be available.
- Replace the "Checkout repository" step with the "Checkout github repo (+ download lfs dependencies)" and "Checkout LFS objects" steps from `.build/workflows/build.yml`.

b4yuan commented:
@pjonsson Updated, let me know what you think!

pjonsson commented:
@b4yuan I think the workflow looks fine, could you please squash your commits so the PR consists of the first commit + a second commit by you?

If you are comparing various code quality tools, you might want to check GCC's `-fanalyzer` switch if you haven't done so already. I think it only catches a few classes of issues so far, but the things I've seen pointed out by it in Contiki-NG have been correct.
